text/0143-sdk-fail-safe-mode.md

```diff
@@ -0,0 +1,512 @@
+
+#### Pros <a name="option-c3-pros"></a>
+
+1. No SDK changes are needed, so it works even for old SDK versions.
+2. This would be a useful feature for our customers even with this RFC.
+
+#### Cons <a name="option-c3-cons"></a>
+
+1. Requires backend changes.
+2. It doesn't work when the SDK starts to crash for a new release for all users, as the backend won't know that there is a new release to expect data from unless users manually create the release.
+
+# Other Ideas to Increase Reliability
+
+This section contains other valuable ideas for increasing SDK reliability that came up during discussions but are out of the scope of this RFC.
+
+## Idea 1: Out of process crash detection <a name="idea-1"></a>
+
+The SDK launches an extra process to monitor and detect a crash in the user’s application. The main advantage is that when the SDK running in the extra process crashes, it doesn’t impact the user’s application process. While this seems appealing, it’s not possible on iOS and Android when writing this, and therefore, we can discard this option.
+
+## Idea 2: Detecting Crashing While Writing a Crash Report <a name="idea-2"></a>
+
+We could detect crashing while writing a crash report by writing a minimal crash report with only bare-bone information and then a full crash report with more details. If the bare-bone crash report misses its counterpart, the complete crash report, we know something is broken.
+
+This has other benefits, too. For example, it would help improve the watchdog termination detection algorithm of the Cocoa SDK.
+
+## Idea 3: Educate Users <a name="idea-3"></a>
+
+Our users could have avoided the [incident mentioned above](#background) by using a staged rollout. While we can't expect all users to use a staged rollout, we could explain actions to increase stability in a section in our docs, helping our users to ship more confidently. We can explain how they can use Sentry in a staged rollout to identify other severe problems, including continuous crashes of any type.
+
+# FAQ
+
+**1. Why do we make the SDK init a NoOp and don't we try to detect which integration is causing the crash?**
+
+We want to keep the logic as simple as possible to reduce the risk of bugs, and it could be that we disable the wrong integration, which would cause more unnecessary crashes.
+
+**2. Why do we focus mostly on crashes during and after SDK init?**
+
+As described in the [motivation](#motivation), continuous SDK crashes during the app start are more critical because they cause the most friction by making the application unusable.
+
+# Drawbacks
+
+Why should we not do this? What are the drawbacks of this RFC or a particular option if
+multiple options are presented:
+
+1. False positives: Potentially wrongly disabling the SDK.
+2. Introducing new crashes with the new logic.
+
+# Unresolved questions
```

Comment:
one question that remained open for me is - do we allow configuration of any of this? to what degree? e.g. number of retries after reverting back from safe mode, etc.
IMO we will definitely want a config option to completely disable all of this, at least in the beginning (will at first probably be opt-in). we'll likely figure out if there's more we want to make configurable along the way